File: basetoolslibrary/src/main/java/com/excellence/basetoolslibrary/utils/FileIOUtils.java
package com.excellence.basetoolslibrary.utils;

import java.io.BufferedOutputStream;
import java.io.BufferedWriter;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.InputStream;
import java.io.OutputStream;

import static com.excellence.basetoolslibrary.utils.EmptyUtils.isEmpty;
import static com.excellence.basetoolslibrary.utils.FileUtils.createNewFile;
import static com.excellence.basetoolslibrary.utils.FileUtils.isFileExists;

/**
 * <pre>
 *     author : VeiZhang
 *     blog   : http://tiimor.cn
 *     time   : 2018/9/10
 *     desc   : 文件流相关
 * </pre> 
 */
public class FileIOUtils {

    private static final int BUF_SIZE = 8 * 1024;

    /**
     * 将字符串写入文件
     *
     * @param file
     * @param content
     * @param append
     * @return
     */
    public static boolean writeFile(File file, String content, boolean append) {
        try {
            if (file == null || isEmpty(content)) {
                return false;
            }
            createNewFile(file);
            BufferedWriter writer = new BufferedWriter(new FileWriter(file, append));
            writer.write(content);
            writer.close();
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * 将字符串写入文件
     *
     * @param path
     * @param content
     * @param append
     * @return
     */
    public static boolean writeFile(String path, String content, boolean append) {
        if (isEmpty(path)) {
            return false;
        }
        return writeFile(new File(path), content, append);
    }

    /**
     * 将字节数组写入文件
     *
     * @param file
     * @param bytes
     * @param append
     * @return
     */
    public static boolean writeFile(File file, byte[] bytes, boolean append) {
        try {
            if (file == null || isEmpty(bytes)) {
                return false;
            }
            createNewFile(file);
            OutputStream os = new BufferedOutputStream(new FileOutputStream(file, append));
            os.write(bytes);
            os.close();
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * 将字节数组写入文件
     *
     * @param path
     * @param bytes
     * @param append
     * @return
     */
    public static boolean writeFile(String path, byte[] bytes, boolean append) {
        if (isEmpty(path)) {
            return false;
        }
        return writeFile(new File(path), bytes, append);
    }

    /**
     * 将输入流写入文件里
     *
     * @param file
     * @param is
     * @param append
     * @param isCloseable 是否关闭输入流
     * @return
     */
    public static boolean writeFile(File file, InputStream is, boolean append, boolean isCloseable) {
        try {
            if (file == null || is == null) {
                return false;
            }
            createNewFile(file);
            OutputStream os = new BufferedOutputStream(new FileOutputStream(file, append));
            byte[] buf = new byte[BUF_SIZE];
            int len;
            while ((len = is.read(buf, 0, buf.length)) != -1) {
                os.write(buf, 0, len);
            }
            os.close();
            if (isCloseable) {
                is.close();
            }
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * 将输入流写入文件里
     *
     * @param path
     * @param is
     * @param append
     * @param isCloseable 是否关闭输入流
     * @return
     */
    public static boolean writeFile(String path, InputStream is, boolean append, boolean isCloseable) {
        if (isEmpty(path)) {
            return false;
        }
        return writeFile(new File(path), is, append, isCloseable);
    }

    /**
     * 读取输入流为字节数组
     *
     * @param is
     * @return
     */
    public static byte[] readFile2Bytes(InputStream is) {
        try {
            ByteArrayOutputStream os = new ByteArrayOutputStream();
            byte[] buf = new byte[BUF_SIZE];
            int len;
            while ((len = is.read(buf, 0, buf.length)) != -1) {
                os.write(buf, 0, len);
            }
            is.close();
            os.close();
            return os.toByteArray();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 读取文件为字节数组
     *
     * @param file
     * @return
     */
    public static byte[] readFile2Bytes(File file) {
        try {
            if (!isFileExists(file)) {
                return null;
            }
            InputStream is = new FileInputStream(file);
            return readFile2Bytes(is);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 读取文件为字节数组
     *
     * @param path
     * @return
     */
    public static byte[] readFile2Bytes(String path) {
        if (isEmpty(path)) {
            return null;
        }
        return readFile2Bytes(new File(path));
    }

    /**
     * 读取输入流为字符串
     *
     * @param is
     * @param charset
     * @return
     */
    public static String readFile2String(InputStream is, String charset) {
        try {
            if (isEmpty(is)) {
                return null;
            }

            byte[] bytes = readFile2Bytes(is);
            if (bytes != null) {
                if (isEmpty(charset)) {
                    return new String(bytes);
                } else {
                    return new String(bytes, charset);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 读取文件为字符串
     *
     * @param file
     * @param charset UTF-8、GBK...
     * @return
     */
    public static String readFile2String(File file, String charset) {
        try {
            if (!isFileExists(file)) {
                return null;
            }
            InputStream is = new FileInputStream(file);
            return readFile2String(is, charset);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return null;
    }

    /**
     * 读取文件为字符串
     *
     * @param path
     * @param charset UTF-8、GBK...
     * @return
     */
    public static String readFile2String(String path, String charset) {
        if (isEmpty(path)) {
            return null;
        }
        return readFile2String(new File(path), charset);
    }

    /**
     * 拷贝文件
     *
     * @param is
     * @param os
     * @return
     */
    public static boolean copyFile(InputStream is, OutputStream os) {
        try {
            if (is == null || os == null) {
                return false;
            }
            byte[] buf = new byte[BUF_SIZE];
            int len;
            while ((len = is.read(buf, 0, buf.length)) != -1) {
                os.write(buf, 0, len);
            }
            is.close();
            os.close();
            return true;
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * 拷贝文件
     *
     * @param sourceFile
     * @param targetFile
     * @return
     */
    public static boolean copyFile(File sourceFile, File targetFile) {
        if (!isFileExists(sourceFile)) {
            return false;
        }
        try {
            return copyFile(new FileInputStream(sourceFile), new FileOutputStream(targetFile));
        } catch (Exception e) {
            e.printStackTrace();
        }
        return false;
    }
}

File: basetoolslibrary/src/main/java/com/excellence/basetoolslibrary/utils/FileIOUtils.kt
package com.excellence.basetoolslibrary.utils

import com.excellence.basetoolslibrary.utils.EmptyUtils.isEmpty
import java.io.*
import java.nio.charset.Charset

/**
 * <pre>
 *     author : VeiZhang
 *     blog   : http://tiimor.cn
 *     time   : 2022/4/7
 *     desc   : 文件流相关
 * </pre>
 */
object FileIOUtils {

    private const val BUF_SIZE = 8 * 1024

    /**
     * 将字符串写入文件
     *
     * @param file
     * @param content
     * @param append
     * @return
     */
    @JvmStatic
    fun writeFile(file: File?, content: String?, append: Boolean): Boolean {
        try {
            if (file == null || isEmpty(content)) {
                return false
            }
            FileUtils.createNewFile(file)
            val writer = BufferedWriter(FileWriter(file, append))
            writer.write(content)
            writer.close()
            return true
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * 将字符串写入文件
     *
     * @param path
     * @param content
     * @param append
     * @return
     */
    @JvmStatic
    fun writeFile(path: String?, content: String?, append: Boolean): Boolean {
        if (isEmpty(path)) {
            return false
        }
        return writeFile(File(path), content, append)
    }

    /**
     * 将字节数组写入文件
     *
     * @param file
     * @param bytes
     * @param append
     * @return
     */
    @JvmStatic
    fun writeFile(file: File?, bytes: ByteArray?, append: Boolean): Boolean {
        try {
            if (file == null || isEmpty(bytes)) {
                return false
            }
            FileUtils.createNewFile(file)
            val os: OutputStream = BufferedOutputStream(FileOutputStream(file, append))
            os.write(bytes)
            os.close()
            return true
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * 将字节数组写入文件
     *
     * @param path
     * @param bytes
     * @param append
     * @return
     */
    @JvmStatic
    fun writeFile(path: String?, bytes: ByteArray?, append: Boolean): Boolean {
        if (isEmpty(path)) {
            return false
        }
        return writeFile(File(path), bytes, append)
    }

    /**
     * 将输入流写入文件里
     *
     * @param file
     * @param is
     * @param append
     * @param isCloseable 是否关闭输入流
     * @return
     */
    @JvmStatic
    fun writeFile(file: File?, stream: InputStream?, append: Boolean, isCloseable: Boolean): Boolean {
        try {
            if (file == null || stream == null) {
                return false
            }

            FileUtils.createNewFile(file)
            val os: OutputStream = BufferedOutputStream(FileOutputStream(file, append))
            val buf = ByteArray(BUF_SIZE)
            var len: Int
            while (stream.read(buf, 0, buf.size).also { len = it } != -1) {
                os.write(buf, 0, len)
            }
            os.close()
            if (isCloseable) {
                stream.close()
            }
            return true
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * 将输入流写入文件里
     *
     * @param path
     * @param stream
     * @param append
     * @param isCloseable 是否关闭输入流
     * @return
     */
    @JvmStatic
    fun writeFile(path: String?, stream: InputStream?, append: Boolean, isCloseable: Boolean): Boolean {
        if (isEmpty(path)) {
            return false
        }
        return writeFile(File(path), stream, append, isCloseable)
    }

    /**
     * 读取输入流为字节数组
     *
     * @param stream
     * @return
     */
    @JvmStatic
    fun readFile2Bytes(stream: InputStream?): ByteArray {
        try {
            val os = ByteArrayOutputStream()
            val buf = ByteArray(BUF_SIZE)
            var len: Int
            while (stream!!.read(buf, 0, buf.size).also { len = it } != -1) {
                os.write(buf, 0, len)
            }
            stream.close()
            os.close()
            return os.toByteArray()
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return ByteArray(0)
    }

    /**
     * 读取文件为字节数组
     *
     * @param file
     * @return
     */
    @JvmStatic
    fun readFile2Bytes(file: File?): ByteArray {
        try {
            if (!FileUtils.isFileExists(file)) {
                return ByteArray(0)
            }
            val stream: InputStream = FileInputStream(file)
            return readFile2Bytes(stream)
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return ByteArray(0)
    }

    /**
     * 读取文件为字节数组
     *
     * @param path
     * @return
     */
    @JvmStatic
    fun readFile2Bytes(path: String?): ByteArray? {
        if (isEmpty(path)) {
            return null
        }
        return readFile2Bytes(File(path))
    }

    /**
     * 读取输入流为字符串
     *
     * @param stream
     * @param charset
     * @return
     */
    @JvmStatic
    fun readFile2String(stream: InputStream?, charset: String?): String {
        try {
            if (isEmpty(stream)) {
                return ""
            }
            val bytes = readFile2Bytes(stream)
            if (isEmpty(charset)) {
                return String(bytes)
            } else {
                return String(bytes, Charset.forName(charset))
            }
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return ""
    }

    /**
     * 读取文件为字符串
     *
     * @param file
     * @param charset UTF-8、GBK...
     * @return
     */
    @JvmStatic
    fun readFile2String(file: File?, charset: String?): String {
        try {
            if (!FileUtils.isFileExists(file)) {
                return ""
            }
            val stream: InputStream = FileInputStream(file)
            return readFile2String(stream, charset)
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return ""
    }

    /**
     * 读取文件为字符串
     *
     * @param path
     * @param charset UTF-8、GBK...
     * @return
     */
    @JvmStatic
    fun readFile2String(path: String?, charset: String?): String {
        if (isEmpty(path)) {
            return ""
        }
        return readFile2String(File(path), charset)
    }

    /**
     * 拷贝文件
     *
     * @param stream
     * @param os
     * @return
     */
    @JvmStatic
    fun copyFile(stream: InputStream?, os: OutputStream?): Boolean {
        try {
            if (stream == null || os == null) {
                return false
            }
            val buf = ByteArray(BUF_SIZE)
            var len: Int
            while (stream.read(buf, 0, buf.size).also { len = it } != -1) {
                os.write(buf, 0, len)
            }
            stream.close()
            os.close()
            return true
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * 拷贝文件
     *
     * @param sourceFile
     * @param targetFile
     * @return
     */
    @JvmStatic
    fun copyFile(sourceFile: File?, targetFile: File?): Boolean {
        if (!FileUtils.isFileExists(sourceFile)) {
            return false
        }
        try {
            return copyFile(FileInputStream(sourceFile), FileOutputStream(targetFile))
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * 拷贝文件
     * 1.起始位置
     * 2.截止长度，最大为原始文件长度；<=0，表示不截取
     *
     * @param sourceFile
     * @param targetFile
     * @param skip
     * @param targetSize
     * @return
     */
    @JvmStatic
    fun copyFile(sourceFile: File, targetFile: File?, skip: Long, targetSize: Long): Boolean {
        if (!FileUtils.isFileExists(sourceFile)) {
            return false
        }
        try {
            val stream: InputStream = FileInputStream(sourceFile)
            val os: OutputStream = FileOutputStream(targetFile)
            stream.skip(skip)

            if (targetSize > 0 && targetSize < sourceFile.length()) {
                if (targetSize <= BUF_SIZE) {
                    val buf = ByteArray(targetSize.toInt())
                    var len: Int
                    if (stream.read(buf, 0, buf.size).also { len = it } != -1) {
                        os.write(buf, 0, len)
                    }
                } else {
                    val buf = ByteArray(BUF_SIZE)
                    var len: Int
                    var count: Long = 0
                    while (stream.read(buf, 0, buf.size).also { len = it } != -1) {
                        os.write(buf, 0, len)
                        count += len.toLong()

                        if (targetSize + BUF_SIZE > count) {
                            len = stream.read(buf, 0, buf.size)
                            os.write(buf, 0, len)
                            count += len.toLong()
                        }

                        if (targetSize <= count) {
                            break
                        }
                    }
                }
                stream.close()
                os.close()
                return true
            } else {
                return copyFile(stream, os)
            }
        } catch (e: java.lang.Exception) {
            e.printStackTrace()
        }
        return false
    }

    /**
     * 拷贝文件，截止长度，最大为原始文件长度；<=0，表示不截取
     *
     * @param sourceFile
     * @param targetFile
     * @param targetSize
     * @return
     */
    @JvmStatic
    fun copyFile(sourceFile: File, targetFile: File?, targetSize: Long): Boolean {
        return copyFile(sourceFile, targetFile, 0, targetSize)
    }
}